### PR TITLE
#sql-task. andrei-zhivotkov

### PR DESCRIPTION
**Pull request by Inquisitor1229**
https://travis-ci.org/github/Inquisitor1229/db-assignments/builds/724741683

### CODE REVIEW

File: task/sql-tasks.js [hunk 1]
```diff
@@ -91,7 +129,18 @@ async function task_1_5(db) {
...
  */
 async function task_1_6(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT 
+            ProductName,
+            CategoryName,
+            CompanyName AS "SupplierCompanyName"
+        FROM Products
+        JOIN Categories ON Products.CategoryID = Categories.CategoryID
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use inner join explicitly. Fix in the whole file not just in this task

File: task/sql-tasks.js [hunk 2]
```diff
@@ -105,7 +154,17 @@ async function task_1_6(db) {
...
  */
 async function task_1_7(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            employee1.EmployeeId AS "EmployeeId",
+            concat( employee1.FirstName, " ", employee1.LastName) AS "FullName",
+            COALESCE(concat( employee2.FirstName, " ", employee2.LastName), "-") AS "ReportsTo"
+        FROM Employees employee2
+        RIGHT JOIN Employees employee1 ON employee2.EmployeeID = employee1.ReportsTo
```
Comment (AlexMokin) [inline task/sql-tasks.js]: do not use right join if not necessary

File: task/sql-tasks.js [hunk 3]
```diff
@@ -129,7 +197,15 @@ async function task_1_8(db) {
  *
  */
 async function task_1_9(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT 
+            CustomerID,
+            ContactName
+        FROM Customers
+        WHERE ContactName LIKE "F%___n%"
```
Comment (AlexMokin) [inline task/sql-tasks.js]: n must be exactly 4rth character

File: task/sql-tasks.js [hunk 4]
```diff
@@ -177,7 +278,14 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+	        COUNT(ProductName) AS "TotalOfCurrentProducts",
+	        SUM(Discontinued) AS "TotalOfDiscontinuedProducts"
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count recrods to show intent explicitly

Comment (AlexMokin) [reply task/sql-tasks.js]: but in this case it's even better to do this 

```
select
(select count()...) as 'T'
(select count()...) as 'TT'
```

notice that from clause is optional

File: task/sql-tasks.js [hunk 5]
```diff
@@ -199,7 +316,25 @@ async function task_1_14(db) {
  *
  */
 async function task_1_15(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
+            SUM(MONTH(OrderDate) = "1") AS "January",
```
Comment (AlexMokin) [inline task/sql-tasks.js]: use count when you need to count recrods to show intent explicitly

File: task/sql-tasks.js [hunk 6]
```diff
@@ -280,7 +475,26 @@ async function task_1_21(db) {
...
+        JOIN Orders ON Customers.CustomerID = Orders.CustomerID
+        JOIN OrderDetails ON Orders.OrderID = OrderDetails.OrderID
+        JOIN Products ON OrderDetails.ProductID = Products.ProductID
+        WHERE OrderDetails.UnitPrice = 
+            (SELECT
+                MAX(orderdetails.UnitPrice)
+            FROM Customers customers1
+            JOIN Orders orders ON orders.CustomerID = customers1.CustomerID
+            JOIN OrderDetails orderdetails ON orderdetails.OrderID = orders.OrderID
+            WHERE Customers.CompanyName = customers1.CompanyName
```
Comment (AlexMokin) [inline task/sql-tasks.js]: company name doesn't uniquely identify customer. also this column doesn't have an index so it's also slow.

Comment (AlexMokin) [reply task/sql-tasks.js]: also try to group by customer id and join subquery on price and customer id

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js [hunk 7]
```diff
@@ -129,7 +197,15 @@ async function task_1_8(db) {
  *
  */
 async function task_1_9(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT 
+            CustomerID,
+            ContactName
+        FROM Customers
+        WHERE ContactName LIKE "F%__n%"
```
Comment (AlexMokin) [inline task/sql-tasks.js]: n must be exactly 4rth character

File: task/sql-tasks.js [hunk 8]
```diff
@@ -177,7 +278,13 @@ async function task_1_12(db) {
  *
  */
 async function task_1_13(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT
```
Comment (AlexMokin) [inline task/sql-tasks.js:282]: In this case it's even better to do this

```
select
(select count()...) as 'T'
(select count()...) as 'TT'
```
notice that from clause is optional

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

File: task/sql-tasks.js (hunk 8 above)
Comment (Inquisitor1229) [reply task/sql-tasks.js:282]: > In this case it's even better to do this
> 
> ```
> select
> (select count()...) as 'T'
> (select count()...) as 'TT'
> ```
> 
> notice that from clause is optional

Could you elaborate on what is wrong here? I have the same structure as you wrote:
```
SELECT
	        (SELECT COUNT(ProductName)) AS "TotalOfCurrentProducts",
	        (SELECT SUM(Discontinued)) AS "TotalOfDiscontinuedProducts"
        FROM Products;
```
Maybe I don't understand something

File: task/sql-tasks.js [hunk 9]
```diff
@@ -129,7 +197,15 @@ async function task_1_8(db) {
  *
  */
 async function task_1_9(db) {
-    throw new Error("Not implemented");
+    let result = await db.query(`
+        SELECT 
+            CustomerID,
+            ContactName
+        FROM Customers
+        WHERE ContactName LIKE "F%_n%"
```
Comment (AlexMokin) [inline task/sql-tasks.js]: n must be exatly 4rth character 
https://www.w3schools.com/sql/sql_like.asp

**Review by AlexMokin — COMMENTED** [review]
Looks good. See comments.
I'll mark the PR as "Draft", please click "ready for review" when it will be finished. Thank you!

**Comment by AlexMokin** [comment]
accepted

**Pull request closed without merging**